1. Allez récupérer sur internet trois photos :   
    - 1 photo de chat que vous enregistrerez sous cat.jpg
    - 1 photo de chien que vous enregistrerez sous dog.jpg
    - 1 photo d'une tasse que vous enregistrerez sous mug.jpg

# Transfer Learning

## Load the VGG16 Pre-trained Model

2. Faites des recherches sur ce modèle VGG16. Depuis quand date ce modèle ? Quelle est son architecture ? Sur quelles données se basent ce modèle ? etc

Ce modèle date de 2014 et à été le premier au classement de la compétition ILSVRC dans la catégorie "classification task".
Son architecture est faite d'un réseau de plusieurs couches de neurones convolutionnels de profondeurs accrues pour améliorer son efficacité avec entre chacun des filtres de poolings (5). Au total on a une pile de 3 couches connectés incluant chacun des chaines comme 4096 et 1000 et c'est seulement la dernière couche qui est une soft-max layer.

Le modèle se base sur des images de 1000 classes différentes triéesen trois jeux de 1.3 million images d'entrainement, 100,000 images de tests and 50,000 images de validation.

In [1]:
# example of loading the vgg16 model
from keras.applications.vgg16 import VGG16
# load model
vgg16_model = VGG16()
# summarize the model
vgg16_model.summary()

2021-12-17 11:06:19.660259: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:305] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
2021-12-17 11:06:19.660354: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:271] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 0 MB memory) -> physical PluggableDevice (device: 0, name: METAL, pci bus id: <undefined>)


Metal device set to: Apple M1 Pro

systemMemory: 16.00 GB
maxCacheSize: 5.33 GB

Model: "vgg16"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 224, 224, 3)]     0         
                                                                 
 block1_conv1 (Conv2D)       (None, 224, 224, 64)      1792      
                                                                 
 block1_conv2 (Conv2D)       (None, 224, 224, 64)      36928     
                                                                 
 block1_pool (MaxPooling2D)  (None, 112, 112, 64)      0         
                                                                 
 block2_conv1 (Conv2D)       (None, 112, 112, 128)     73856     
                                                                 
 block2_conv2 (Conv2D)       (None, 112, 112, 128)     147584    
                                              

3. Globalement les couches utilisées ne sont pas nouvelles et ont été vues sur le dernier TP, excepté MaxPooling2D : quel est son rôle ?

Cette couche permet de réduire la taille des pixel d'une image, plus on augmente sa valeur plus on perdra en qualité et en précision mais on gagnera en vitesse de traitement du modèle.

# Load the ResNet50 Pre-trained Model

4. Faites des recherches sur ce modèle ResNet50. Depuis quand date ce modèle ? Quelle est son architecture ? etc

In [2]:
# example of loading the resnet50 model
from keras.applications.resnet import ResNet50
# load model
resnet_model = ResNet50()
# summarize the model
resnet_model.summary()

Model: "resnet50"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_2 (InputLayer)           [(None, 224, 224, 3  0           []                               
                                )]                                                                
                                                                                                  
 conv1_pad (ZeroPadding2D)      (None, 230, 230, 3)  0           ['input_2[0][0]']                
                                                                                                  
 conv1_conv (Conv2D)            (None, 112, 112, 64  9472        ['conv1_pad[0][0]']              
                                )                                                                 
                                                                                           

5. Idem, on connait la majorité des couches, exceptées :   
    ZeroPadding2D  
    BatchNomalization  
Que pouvez-vous en dire ?

ZeroPadding2D : Cette couche peut ajouter des rangées et des colonnes de zéros en haut, en bas, à gauche et à droite d'une image.

Batch normalization : Normalise ses entrées par lots en appliquant une transformation qui maintient la sortie moyenne proche de 0 et l'écart type de la sortie proche de 1.


# Prédiction

Vous allez tester le code ci-dessous avec chacune des photos (cat, dog et mug) et les deux modèles (VGG16 et ResNet50 et les prédictions associées).

In [3]:
from keras.applications.vgg16 import preprocess_input, decode_predictions
from keras.preprocessing.image import load_img
from keras.preprocessing.image import img_to_array

images = []
for img in ["dog", "cat", "mug"]:
    # load an image from file
    image = load_img("%s.jpg" % img, target_size=(224, 224))
    # convert the image pixels to a numpy array
    image = img_to_array(image)
    # reshape data for the model
    image = image.reshape((1, image.shape[0], image.shape[1], image.shape[2]))
    print(image)

    images.append(image)

[[[[255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]
   ...
   [255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]]

  [[255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]
   ...
   [255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]]

  [[255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]
   ...
   [255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]]

  ...

  [[255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]
   ...
   [255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]]

  [[255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]
   ...
   [255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]]

  [[255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]
   ...
   [255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]]]]
[[[[255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]
   ...
   [255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]]

  [[255. 255. 255.]
   [255. 255. 255.]
   [255. 255. 255.]
   ...
   [255. 255. 255.]
  

6. Comment a été redimensionné l'image d'origine ?

Elle à été redimensionné grâce au reshape de keras pour avoir une taille conforme à ce que le modèle attend

In [4]:
print("vgg16\n")
resnet_result = []
for image in images:
    # prepare the image for the VGG model
    image = preprocess_input(image)
    # predict the probability across all output classes
    yhat = vgg16_model.predict(image)
    # convert the probabilities to class labels
    label = decode_predictions(yhat)
    # retrieve the most likely result, e.g. highest probability
    label = label[0][0]
    resnet_result.append(label)

vgg16



2021-12-17 11:06:22.208675: W tensorflow/core/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
2021-12-17 11:06:22.258169: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:112] Plugin optimizer for device_type GPU is enabled.


In [5]:
# Resnet init
from keras.applications.resnet import ResNet50, preprocess_input, decode_predictions

In [6]:
print('Resnet\n')
print("vgg16\n")
vgg16_result = []
for image in images:
    # prepare the image for the VGG model
    image = preprocess_input(image)
    # predict the probability across all output classes
    yhat = resnet_model.predict(image)
    # convert the probabilities to class labels
    label = decode_predictions(yhat)
    # retrieve the most likely result, e.g. highest probability
    label = label[0][0]
    vgg16_result.append(label)

Resnet

vgg16



2021-12-17 11:06:22.818954: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:112] Plugin optimizer for device_type GPU is enabled.


In [7]:
print(resnet_result)
print(vgg16_result)


[('n02099601', 'golden_retriever', 0.4258668), ('n02127052', 'lynx', 0.6526927), ('n07930864', 'cup', 0.5259743)]
[('n02102318', 'cocker_spaniel', 0.7528526), ('n02127052', 'lynx', 0.27837187), ('n03063599', 'coffee_mug', 0.6520036)]


* Chaque % = taux de confiance des prévision du modèle
|image|vgg|resnet|
|----|----|------|
|dog|75%|42%|
|cat|27%|65%|
|cup|65%|50%|

**Vgg performe mieux que restnet sur le chien et le mug.**


7. Faites un tableau récapitulatif des prédictions pour chacune des images selon le modèle, ainsi que la pourcentage de confiance associé.